### PR TITLE
fix(core): make UPGD gradient alignment scale-free across underflow and overflow (#2389)

### PR DESCRIPTION
**Pull request by Kayvan-Zahiri**
Fixes #2389.

## Summary
In `UPGDLearner`, `_tuple_norm` and `_gradient_alignment` used unscaled sums of squares and hardcoded `1e-6` / `1e-12` floors. For gradient norms below `1e-6` (or elements underflowing in float32 squared norms below `1e-20`), the alignment collapsed to `0.0`, disabling meta-plasticity updates. For large gradients (above `1e+10`), unscaled sums of squares overflowed to `nan`.

## Proposed Changes
1. Used exact power-of-two rescaling (`frexp` and `ldexp`) in `_tuple_norm` and `_gradient_alignment` to make vector norms and cosine calculations scale-free across float32 dynamic range.
2. Preserved the exact reduction to returning `0.0` for all-zero or non-finite gradient inputs.
3. Added unit tests in `tests/test_upgd.py` verifying that identical and opposing gradient tuples consistently produce alignments of `+1.0` and `-1.0` across scales from `1e-20` up to `1e+10`.

## Test Plan
- `uv run pytest tests/test_upgd.py` (89 passed)
- `uv run ruff check alberta_framework/core/upgd.py tests/test_upgd.py` (clean)
- `uv run mypy alberta_framework/core/upgd.py` (clean)

### CODE REVIEW

**Review by agent-cortex — APPROVED** [review]
## Review — APPROVE

I reproduced the numerical instability this PR targets and verified the rewritten `_gradient_alignment` fixes it at every scale, then traced every caller. The shipped behavior is correct; two non-blocking items below.

### Scope
`+54/−9` across `alberta_framework/core/upgd.py` (`_tuple_norm`, `_gradient_alignment`) and `tests/test_upgd.py`. No unrelated behavior.

### The instability is real (reproduced, both directions)
Base `_gradient_alignment` computed `dot / (prev_norm * curr_norm + 1e-12)` behind a `(prev_norm > 1e-6) & (curr_norm > 1e-6)` gate, with `_tuple_norm = sqrt(sum(x^2) + 1e-12)`. Both the `1e-12` additive floor and the `1e-6` gate are absolute constants, so the result is **not** scale-invariant. I replicated the exact base vs. head logic in float32 and ran the PR's own fixture (`g1` scaled by `10**e`), comparing identical gradients (should be `+1`) and opposed gradients (should be `-1`):

```
  exp    base_same  head_same     base_opp   head_opp
   10            1          1           -1         -1
    4            1          1           -1         -1
    0            1          1           -1         -1
   -4      0.99999          1     -0.99999         -1
   -8   0.00076192          1  -0.00076192         -1     <- 1e-12 floor swamps the norm
  -10            0          1            0         -1     <- 1e-6 gate returns 0
  -20            0          1            0         -1
```

And the overflow half of the title's claim:

```
exp 18: base_same=1     head_same=1
exp 20: base_same=nan   head_same=1     <- x^2 overflows float32 -> inf/inf
exp 25: base_same=nan   head_same=1
```

So on `main`, cosine alignment of *identical* gradients degrades to `0.00076`, then `0.0` under downscaling, and to `nan` under upscaling — exactly the failure the PR describes. The head rescales each tuple by a power of two from `frexp(max_abs)` before accumulating, so the `2^-exp` factors cancel in `dot/(p_norm*c_norm)` and the cosine is exact `±1` from `10^-20` to `10^25`. I confirmed the edge cases too: empty tuples and all-zero gradients hit `active=False` and return `0.0` (matching the docstring), and non-finite `cos` is caught by `valid = active & isfinite(cos)`.

The new `test_gradient_alignment_scale_invariance_underflow_overflow` therefore bites: on `main` the `atol=1e-5` assertion fails hard (`0.0`/`0.00076` vs `1.0`) and passes at head.

### Non-blocking: the rewritten `_tuple_norm` and `_tuple_dot` are now dead code
`_gradient_alignment` was the sole consumer of both helpers, and the head inlines its own scaled `dot`/`p_sq`/`c_sq` accumulation — it no longer calls `_tuple_norm` or `_tuple_dot`. Grepping the whole tree confirms it:

```
$ grep -rn "_tuple_norm\|_tuple_dot" --include=*.py .
alberta_framework/core/upgd.py:1978:    def _tuple_dot(...):
alberta_framework/core/upgd.py:1986:    def _tuple_norm(...):
```

i.e. both now appear only at their own definitions. So this PR spent ~10 lines rewriting `_tuple_norm` for scale-freeness that nothing reaches (the rewrite is untested/unreachable), and `_tuple_dot` is left orphaned. Recommend either deleting both, or having `_gradient_alignment` reuse the (now scale-free) `_tuple_norm`/`_tuple_dot` so the helper rewrite is actually exercised. Purely a cleanliness/maintainability point — no runtime effect.

### Non-blocking: the overflow path is correct but untested
The added test only downscales (max exponent `10`), so the `nan`-on-overflow behavior shown above is fixed but not guarded by a test. Since the PR title and summary both claim "overflow," adding one upscaled case (e.g. `exponent = 20`) would pin the half of the claim currently unexercised.

### Transparency on execution
Repo-native execution of `tests/test_upgd.py` was **blocked by a toolchain gap, not skipped**: the sandbox image ships only Python 3.11.2 while `alberta_framework` uses PEP 695 generic syntax (`def _require_host_discount[T](...)`) requiring ≥3.12, and `jaxlib`'s native extension additionally could not `mmap` under the sandbox's `noexec` tmp. I therefore validated the numerics with a faithful float32 replication of the exact base/head logic (shown above) and cross-checked against this PR's CI, whose `runtime py3.12`/`py3.13` shards are green — independent evidence the real test passes at head `562ce2e`.

Correct, well-reproduced numerical-stability fix. Approving; please consider removing the now-dead `_tuple_norm`/`_tuple_dot` and adding an overflow case to the test.

---
AI provider/model: anthropic / claude-opus-4-8
Client / agent tooling: pi-coding-agent
Contribution skill revision: elizaOS/eliza@9b93ec4374878428a7a53b24202dfa15a8dbb55b:packages/skills/skills/contribute-to-eliza
Attribution status: self-reported
— [eliza-swarm]

**Review by kenjohnscreates — COMMENTED** [review]
## Disposition: **close as a later duplicate of #2422** (same rescaled-dot construction; bit-exact family)

Exact head reviewed: `562ce2ef` on base `c9aba7b5`. Filed 2026-09-01 17:30 UTC; #2422 (same author's construction, filed 2026-08-24 18:44 UTC) is the earliest head with this form.

This head is correct on every probe class: the scale sweep, near-float32-max and dominant-entry rows, `finfo.tiny`, `+-inf` entries, all-zero / empty tuples and all end-to-end rows match #2422 row for row, and the 3000-row well-formed sweep is byte-identical to main (`1e23257331cb`), the property that separates this family from the unit-vector heads. It also rewrites `_tuple_norm` with the rescale, zips `previous` and `current` (fine here since the tuples are always parallel), and clips to `[-1, 1]`, which makes its 80-step meta hash match #2519 rather than #2422. One measured edge: a zero-size array inside a tuple raises `ValueError` from `jnp.max` without `initial=`; main and #2422 return `0.0`. That input is unreachable through the public constructor (layer widths are validated `>= 1`), so it is a note, not a defect. Own `tests/test_upgd.py` passes (89); ruff clean.

Under the cluster's one-approval discipline the approval goes to #2422.

## Measured across the cluster

Three constructions:

- **rescaled-dot** (#2422, #2465, #2507, #2519, #2771): `ldexp` each tuple by `-frexp(max|x|)`, take the dot on the rescaled copies and divide by the product of the rescaled norms;
- **unit-vector** (#2390, #2530, #2733, #2752): same rescale, then divide each element by its rescaled norm and dot the unit tuples;
- **peak-division** (#2444): divide each tuple by `max|entry|` (not a power of two), then the original `_tuple_norm` with its `1e-12` term.

| Probe row | main | rescaled-dot | unit-vector | #2444 peak-division |
|---|---|---|---|---|
| `cos(g,g)`, `cos(g,-g)` at 1e-10 ... 1e-30 | `0.0`, `0.0` | `1.0` / `-1.0` (exact or 1 ulp) | `0.99999982` ... `1.0000002` | `0.99999994` |
| same at 1e19 | `nan` | `1.0` / `-1.0` | `0.99999994` | `0.99999994` |
| `cos(g,g)` at 1e-38 and 1e-44 (below `finfo.tiny`, squares flush) | `0.0` | `0.0` | `0.0` | `0.0` |
| `[3e38,-1.5e38,7.5e37]` self / anti | `nan` | `1.0` / `-1.0` | `1.0` / `-1.0` | **`0.0` / `0.0`** |
| `[1e38,1e-8,0]` self; vs `[1e38,0,0]` | `nan` | `1.0`; `1.0` | `1.0`; `1.0` | **`0.0`; `0.0`** |
| `[finfo.max]*2` self | `nan` | `1.0` | `0.99999994` | **`0.0`** |
| `[finfo.tiny]*2` self | `0.0` | `1.0` | `0.99999994` | `1.0` |
| `+-inf` entry, either side | `nan` | `0.0` | `0.0` | `nan` (#2444: `nan`) |
| all-zero / empty tuple | `0.0` | `0.0` | `0.0` | `0.0` |
| zero-size array in tuple | `0.0` | #2422 #2465: `0.0`; #2507 #2519 #2771: **ValueError** | #2390 #2733: `0.0`; #2530 #2752: **ValueError** | `0.0` |
| 3000-row well-formed O(1) bit sweep | `1e23257331cb` | **`1e23257331cb` (bit-identical)** | `811757cf5f0b` | `43b93e00fff2` |
| 1000-row power-of-two-scaled sweep | `cf070249d273` (contains 0.0/nan rows) | `a02828911586` | `fad5b8cca31c` | `71ac19d8d3a6` |
| end-to-end `meta_head_weight_log_scale` at 1e-4 ... 1e-10 (want +0.5) | `2.8e-5`, `0.0`, ... | `0.5` (+-3e-8) | `0.5` (+-3e-8) | `0.5` |
| same at 1e10 (want +0.5) | `nan` | `0.5` | `0.5` | `0.5` |
| reversed gradient at 1e-6, 1e-10 (want -0.5) | `0.0` | `-0.5` | `-0.5` | `-0.5` |
| 80-step learner hash, `adaptive_kappa_mode=gradient_alignment`, 3 scales | ref | identical to main (all heads) | identical | identical |
| 80-step learner hash, `meta_plasticity_mode=gradient_alignment` | ref | #2422 = #2465 = #2507; #2519 = #2771 (they clip) | #2390 = #2733; #2530, #2752 distinct | distinct |

The meta-plasticity 80-step hashes differ from main on every head because main's `1e-12`-inside-`sqrt` floor already touches O(1) per-parameter gradients in this 4-8-2 net, so that path is not a bit-exactness criterion; the adaptive-kappa path and the 3000-row sweep are. The rescaled-dot construction is the one that leaves every well-formed value byte-identical to main (power-of-two scaling commutes with rounding, and the common factor cancels in the ratio); forming unit vectors first re-rounds every entry, which is why the unit-vector heads report `cos(g,g)` as `0.99999982` rather than `1.0`. #2444 divides by a non-power-of-two peak, and for entries above ~1e38 the float32 reciprocal of that divisor is subnormal and flushes, so the whole tuple becomes zero and the row returns `0.0` -- exactly the failure the issue text warned against. Every head passes its own `tests/test_upgd.py` and `ruff` (except #2465: one E501).


---
Reviewed with claude-code / anthropic / claude-fable-5. arm64 macOS, CPython 3.13, JAX 0.11.0 (XLA:CPU), float32 defaults.

**Comment by lalalune** [comment]
Closing as superseded by merged #2733. The merged power-of-two implementation handles empty leaves and nonfinite directions and verifies live meta-plasticity changes. Four regressions fail on main and all 93 UPGD tests pass; direct reciprocal scaling alternatives can lose finite extreme gradients.
